### ISSUE TITLE
Fix false positives for maps in value-keyword-case

### ISSUE THREAD

**Issue by MrVladevoit** [stylelint/stylelint#4882]
> Clearly describe the bug

Rule value-keyword-case highlight syntax inside [LESS lookups](http://lesscss.org/features/#maps-feature-using-variable-variables-in-lookups) in [LESS maps](http://lesscss.org/features/#maps-feature). 

> Which rule, if any, is the bug related to?

value-keyword-case

> What code is needed to reproduce the bug?

e.g.

```LESS
@map: {
  key: valueName1 valueName2;
}
```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "rules": {
    "value-keyword-case": "lower",
  }
}
```

> Which version of stylelint are you using?

13.6.1

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

"Webstorm plugin"

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

"Yes, it's related to LESS maps."

> What did you expect to happen?

"No warnings inside LESS maps syntaxes."

> What actually happened (e.g. what warnings or errors did you get)?

 Expected "valueName1" to be valuename1

**Comment by malsf21** [comment]
Hi @MrVladevoit,

Thank you for using the issue template! I can the behaviour that you listed [with this demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIACBbAhgB3gAmAHQDttsBrGATxwDc0AbAVxgDk0UYBGbGh51mAJgDchAL4gANCADO5AgBc0ADwQhaMKVIngIBAGYBLAOYr8RbHhAAnemqkWcp4sQvdGAWjLkA7hEtQ3YGhSMPbmqhBeMJYWhMQioiAiQA).

However, I believe the rule is working as intended. [As this entry in the docs says](https://stylelint.io/user-guide/rules/value-keyword-case),

> Keyword values which are paired with non-properties (e.g. `$vars` and custom properties), and do not conform to the primary option, can be ignored using the `ignoreKeywords: []` secondary option.

And in this case, `key` is a custom property to LESS, with an infinite set of keyword values. If you know you will only use a certain set of variables, you could use `ignoreKeywords: []` on that set of variables.

If you would like instead, you could also use [`ignoreProperties: []`](https://stylelint.io/user-guide/rules/value-keyword-case#ignoreproperties-regex-regex-non-regex) option on `value-keyword-case` and select `key` as one of your properties; that should resolve the problem.

That being said, there may be something that we can do to resolve this issue and make it more convenient for LESS developers. I'll tag this as needs discussion in the meantime, but let me know if the comments made sense!

**Comment by MrVladevoit** [comment]
@malsf21 , thanks for your answer ! 

The LESS lookups it's not a keywords or CSS values. I don't understand why i must add this values to ignoreKeywords: []. 

If i write lookups outside of LESS maps stylelint will not to warning me.

```less 
@set: blueColor greenColor redColor;

each(@set{
   .color2--@{value} {
     color: @@value;
     }
});
```

But if i write lookups inside LESS maps it will be error: 

```less
@blueColor: blue;
@greenColor: green;
@redColor: red;
@yellowColor: yellow;
@orangeColor: orange;
@blackColor: black;
@whiteColor: white;

@set: {
  one: blueColor greenColor redColor;
  two: yellowColor orangeColor;
  three: whiteColor blackColor;
}

each(@set[one]{
   .color--@{value} {
     color: @@value;
     }
  });
```

See [demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIACAjANgVxgYQmiATvAATpYDcAOgHZIDmBMM1u+RxDT1VtjUrhEnx5IAnjDT4A7gPbjJEKSMIBDanRx5BxVepgj0KsAGtZJQyZFSAFgEsALprYkbD-TRpIAzjHslgNMQ61DDmmE6EHIzMssR8sjxB9lIQJPLSsboaCYHE9tbRLnaOsRamWgQ8AL65nj72AExhWLGcMRVxMPwVPIEwRtYAFN6+ANoQIQC6AdRBxAB0kGwAtMtIwABuKuFVxDNzQUvaSEhb4YkHNbPEVQCUvbP9YEMjjftBixUNq+tnWLvvOZHdgnP7ua5zK5BO6JEAAGhAXlE1HsKgAHggQGgYF4vPDwBMAGa2OiY96UEAEDDYrwU-y5IIUsHLYwwUQpAhQZZgFQ+OnECnSGAECm5K5VEBVIA)

I think that stylelint cannot parse LESS maps correctly.

**Comment by MrVladevoit** [comment]
You can read about [Using variable variables in lookups ](http://lesscss.org/features/#maps-feature-using-variable-variables-in-lookups)

**Comment by malsf21** [comment]
Hey @MrVladevoit,

Thanks for the response! I was probably wrong then, I'm not familiar with LESS (thank you for providing the links), and I do agree that the behaviour now seems slightly strange. I think [this demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIYEMwAsAUABAzjALvAATAA6AdkUROTMQEIA2AgmANIDcFAvtwJRdyFXAWJlK1WgxbtB3EABoQOAJ7l8yAB4IQjGDhyLwNAGYBLAOY7xVUiABOAVz047YilVsgAbskaOYAFoAaxgVAHcIeyhAsGQ8NyI7RghwmHs7DyJuHhBuIA) (which isn't exactly semantic LESS, though that's not important) is probably the most telling comparison. 

I'm not very familiar with the LESS parsing, but I assume this has to do with how we're parsing the `each` and the `@set` items: it may be that `@set` is incorrectly being parsed as an at-rule, or that `each` is designed to return early (since the stuff inside might be a variable lookup, as you've shown). I may be able to take a crack at this later, but I encourage you to poke around the files if you have a chance.

I'll re-label this as needs investigation and help wanted for now. Thank you for taking the time to report this properly!

**Comment by jeddy3** [comment]
The rule should ignore LESS maps. 

I suggest we:

- use the [`isStandardSyntaxDeclaration` util](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxDeclaration.js) in the rule
- tweak the util ([around line 22](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxDeclaration.js#L22-L31)) to return false for LESS maps, i.e. `decl` nodes with a parent node who's selector starts with `@` and ends with `:`.

We use the `isStandard*` utils to ensure the built-in rules only check standard CSS constructs. Here we have a key & value pair of a LESS map masquerading as a declaration. 

I've labelled the issue as ready to implement. @MrVladevoit, please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by fraudoudou** [comment]
I can reproduce the issue for SASS maps as well:
```
$font-faces: (
    default: (
        myFontBold: '...',
        myFontLight: '...'
    ),
    greek: (
        unicode: U+0370-03FF,
        myFontBold: '...',
        myFontLight: '...'
    ),
);
```

All the keys in the map are lower-cased with autofix on and it does not happen if I disable value-keyword-case.

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and fix this bug.
